File: lib/faye/engines/memory.rb
module Faye
  module Engine

    class Memory
      include Timeouts

      def self.create(server, options)
        new(server, options)
      end

      def initialize(server, options)
        @server    = server
        @options   = options
        reset
      end

      def disconnect
        reset
        remove_all_timeouts
      end

      def reset
        @namespace = Namespace.new
        @clients   = {}
        @channels  = {}
        @messages  = {}
      end

      def create_client(&callback)
        client_id = @namespace.generate
        @server.debug('Created new client ?', client_id)
        ping(client_id)
        @server.trigger(:handshake, client_id)
        callback.call(client_id)
      end

      def destroy_client(client_id, &callback)
        return unless @namespace.exists?(client_id)

        if @clients.has_key?(client_id)
          @clients[client_id].each { |channel| unsubscribe(client_id, channel) }
        end

        remove_timeout(client_id)
        @namespace.release(client_id)
        @messages.delete(client_id)
        @server.debug('Destroyed client ?', client_id)
        @server.trigger(:disconnect, client_id)
        @server.trigger(:close, client_id)
        callback.call if callback
      end

      def client_exists(client_id, &callback)
        callback.call(@namespace.exists?(client_id))
      end

      def ping(client_id)
        timeout = @server.timeout
        return unless Numeric === timeout
        @server.debug('Ping ?, ?', client_id, timeout)
        remove_timeout(client_id)
        add_timeout(client_id, 2 * timeout) { destroy_client(client_id) }
      end

      def subscribe(client_id, channel, &callback)
        @clients[client_id] ||= Set.new
        should_trigger = @clients[client_id].add?(channel)

        @channels[channel] ||= Set.new
        @channels[channel].add(client_id)

        @server.debug('Subscribed client ? to channel ?', client_id, channel)
        @server.trigger(:subscribe, client_id, channel) if should_trigger
        callback.call(true) if callback
      end

      def unsubscribe(client_id, channel, &callback)
        if @clients.has_key?(client_id)
          should_trigger = @clients[client_id].delete?(channel)
          @clients.delete(client_id) if @clients[client_id].empty?
        end

        if @channels.has_key?(channel)
          @channels[channel].delete(client_id)
          @channels.delete(channel) if @channels[channel].empty?
        end

        @server.debug('Unsubscribed client ? from channel ?', client_id, channel)
        @server.trigger(:unsubscribe, client_id, channel) if should_trigger
        callback.call(true) if callback
      end

      def publish(message, channels)
        @server.debug('Publishing message ?', message)

        clients = Set.new

        channels.each do |channel|
          next unless subs = @channels[channel]
          subs.each(&clients.method(:add))
        end

        clients.each do |client_id|
          @server.debug('Queueing for client ?: ?', client_id, message)
          @messages[client_id] ||= []
          @messages[client_id] << Faye.copy_object(message)
          empty_queue(client_id)
        end

        @server.trigger(:publish, message['clientId'], message['channel'], message['data'])
      end

      def empty_queue(client_id)
        return unless @server.has_connection?(client_id)
        @server.deliver(client_id, @messages[client_id])
        @messages.delete(client_id)
      end
    end

  end
end

File: src/engines/memory.js
'use strict';

var copyObject = require('../util/copy_object'),
    assign     = require('../util/assign'),
    Namespace  = require('../util/namespace'),
    Set        = require('../util/set'),
    Timeouts   = require('../mixins/timeouts');

var Memory = function(server, options) {
  this._server    = server;
  this._options   = options || {};
  this.reset();
};

Memory.create = function(server, options) {
  return new Memory(server, options);
};

Memory.prototype = {
  disconnect: function() {
    this.reset();
    this.removeAllTimeouts();
  },

  reset: function() {
    this._namespace = new Namespace();
    this._clients   = {};
    this._channels  = {};
    this._messages  = {};
  },

  createClient: function(callback, context) {
    var clientId = this._namespace.generate();
    this._server.debug('Created new client ?', clientId);
    this.ping(clientId);
    this._server.trigger('handshake', clientId);
    callback.call(context, clientId);
  },

  destroyClient: function(clientId, callback, context) {
    if (!this._namespace.exists(clientId)) return;
    var clients = this._clients;

    if (clients[clientId])
      clients[clientId].forEach(function(channel) { this.unsubscribe(clientId, channel) }, this);

    this.removeTimeout(clientId);
    this._namespace.release(clientId);
    delete this._messages[clientId];
    this._server.debug('Destroyed client ?', clientId);
    this._server.trigger('disconnect', clientId);
    this._server.trigger('close', clientId);
    if (callback) callback.call(context);
  },

  clientExists: function(clientId, callback, context) {
    callback.call(context, this._namespace.exists(clientId));
  },

  ping: function(clientId) {
    var timeout = this._server.timeout;
    if (typeof timeout !== 'number') return;

    this._server.debug('Ping ?, ?', clientId, timeout);
    this.removeTimeout(clientId);
    this.addTimeout(clientId, 2 * timeout, function() {
      this.destroyClient(clientId);
    }, this);
  },

  subscribe: function(clientId, channel, callback, context) {
    var clients = this._clients, channels = this._channels;

    clients[clientId] = clients[clientId] || new Set();
    var trigger = clients[clientId].add(channel);

    channels[channel] = channels[channel] || new Set();
    channels[channel].add(clientId);

    this._server.debug('Subscribed client ? to channel ?', clientId, channel);
    if (trigger) this._server.trigger('subscribe', clientId, channel);
    if (callback) callback.call(context, true);
  },

  unsubscribe: function(clientId, channel, callback, context) {
    var clients  = this._clients,
        channels = this._channels,
        trigger  = false;

    if (clients[clientId]) {
      trigger = clients[clientId].remove(channel);
      if (clients[clientId].isEmpty()) delete clients[clientId];
    }

    if (channels[channel]) {
      channels[channel].remove(clientId);
      if (channels[channel].isEmpty()) delete channels[channel];
    }

    this._server.debug('Unsubscribed client ? from channel ?', clientId, channel);
    if (trigger) this._server.trigger('unsubscribe', clientId, channel);
    if (callback) callback.call(context, true);
  },

  publish: function(message, channels) {
    this._server.debug('Publishing message ?', message);

    var messages = this._messages,
        clients  = new Set(),
        subs;

    for (var i = 0, n = channels.length; i < n; i++) {
      subs = this._channels[channels[i]];
      if (!subs) continue;
      subs.forEach(clients.add, clients);
    }

    clients.forEach(function(clientId) {
      this._server.debug('Queueing for client ?: ?', clientId, message);
      messages[clientId] = messages[clientId] || [];
      messages[clientId].push(copyObject(message));
      this.emptyQueue(clientId);
    }, this);

    this._server.trigger('publish', message.clientId, message.channel, message.data);
  },

  emptyQueue: function(clientId) {
    if (!this._server.hasConnection(clientId)) return;
    this._server.deliver(clientId, this._messages[clientId]);
    delete this._messages[clientId];
  }
};

assign(Memory.prototype, Timeouts);

module.exports = Memory;
